bin/massSpecPeptides.pl

```diff
@@ -0,0 +1,432 @@
+  # otherwise we need to try to map all peptides to this protein
+  unless($bestMatch) {
+
+    $peptideLocations = &mapPeptidesToProtein($seqString, $peptideSequencesToRecords);
+    $matchingRecords = &getRecordsFromPeptideSubset($peptideLocations, $peptideSequencesToRecords);
+
+    # NOTE: here we require 100% peptide matches
+    $bestMatch = &findBestRecord($matchingRecords, $peptideLocations, 100);
+  }
+
+  return &cloneAndAddPeptideLocations($bestMatch, $peptideLocations);
+}
+
+
+sub cloneAndAddPeptideLocations {
+  my ($record, $peptideLocations) = @_;
+
+  return unless($record);
+
+  # make a copy of the MSRecord
+  my $clone = $record->clone();
+
+  my $peptideCount = scalar @{$clone->{peptides}};
+
+  foreach my $peptide (@{$clone->{peptides}}) {
+    my $peptideSequence = $peptide->get("sequence");
+
+    foreach my $location (@{$peptideLocations->{$peptideSequence}}) {
+      push @{$clone->{peptideLocations}->{$peptideSequence}}, $location;
+    }
+  }
+  return $clone;
+}
+
+
+sub findBestRecord {
+  my ($matchingRecords, $peptideLocations, $minPeptidePct) = @_;
+
+  my @results;
+
+  foreach my $record (@$matchingRecords) {
+    my $count = 0;
+    my $peptideCount = scalar @{$record->{peptides}};
+
+    foreach my $peptide (@{$record->{peptides}}) {
+      my $peptideSequence = $peptide->get("sequence");
+
+      $count++ if($peptideLocations->{$peptideSequence});
+    }
+
+    my $matchPct = ($count / $peptideCount) * 100;
+
+    if($matchPct >= $minPeptidePct) {
+
+      # NOTE: Use count here, not the pct
+      push @results, [$count, $record];
+    }
+  }
+
+  return unless scalar @results > 0;
+
+  my @sorted = sort { $b->[0] <=> $a->[0]} @results;
+
+
+  # NOTE:  we are returning the record which has the highest
+  # number of peptides matching above our minPeptidePct threshold
+  return $sorted[0]->[1];
+}
+
+
+sub getRecordsFromPeptideSubset {
+  my ($peptideLocations, $peptideSequencesToRecords) = @_;
+
+  my (@res, %seen);
+
+  foreach my $peptide (keys %$peptideLocations) {
+    foreach my $record (@{$peptideSequencesToRecords->{$peptide}}) {
+
+      my $uid = $record->{_unique_id};
+      
+      # NOTE: multiple peptides will return the same record reference
+      # we only want it one time in our array
+      push @res, $record unless($seen{$uid});
+
+      $seen{$uid} = 1;
+
+    }
+  }
+  return \@res;
+}
+
+sub mapPeptidesToProtein {
+  my ($seqString, $peptides) = @_;
+
+  my %res;
+
+  foreach my $peptide (keys %$peptides) {
+    my $pattern = quotemeta($peptide);
+
+    my $pepLength = length $peptide;
+
+    while ($seqString =~ /$pattern/g) {
+      my $end = pos($seqString) + 0;
+      my $start = $end - $pepLength + 1;
+
+      push @{$res{$peptide}}, [$start, $end];
+    }
+  }
+
+  return \%res;
+}
+
+sub getPeptidesFromRecords {
+  my ($records) = @_;
+
+  my %peptides;
+
+  foreach my $record (@$records) {
+    foreach my $peptide ({$record->{peptides}}) {
```

Comment:
```suggestion
    foreach my $peptide (@{$record->{peptides}}) {
```

Doesn't work without this!